File: Leetcode/dfs_recursion_backtracking/79_WordSearch/Solution.java
class Solution {
    public boolean exist(char[][] board, String word) {
        boolean[][] visited = new boolean[board.length][board[0].length];
        boolean res = false;
        for(int i = 0; i<board.length; ++i){
            for(int j = 0; j<board[i].length; ++j){
                if(board[i][j]==word.charAt(0))
                    res |= dfs(board, visited, i, j, word);
            }
        }
        return res;
    }
    private boolean dfs(char[][] board, boolean[][] visited,
                       int i, int j, String pattern){
        if(pattern.length()==0)
            return true;
        if(i<0 || j<0 || i>=board.length || j>=board[i].length || visited[i][j] || pattern.charAt(0)!=board[i][j])
            return false;
        visited[i][j] = true;
        boolean res = dfs(board, visited, i+1, j, pattern.substring(1)) || dfs(board, visited, i-1, j, pattern.substring(1)) ||
                    dfs(board, visited, i, j+1, pattern.substring(1)) || dfs(board, visited, i, j-1, pattern.substring(1));
        visited[i][j] = false;
        return res;
    }
}

File: Leetcode/dfs_recursion_backtracking/79_WordSearch/Solution.py
class Solution(object):
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """
        if not word:
            return True
        visited = [[False for _ in range(len(board[0]))] for _ in range(len(board))]
        def dfs(i, j, pattern):
            if not pattern:
                return True
            if i<0 or j<0 or i>=len(board) or j>=len(board[i]) or visited[i][j] or pattern[0]!=board[i][j]:
                return False
            visited[i][j] = True
            res = dfs(i+1, j, pattern[1:]) or dfs(i, j+1, pattern[1:]) or dfs(i-1, j, pattern[1:]) or dfs(i, j-1, pattern[1:])
            visited[i][j] = False
            return res
        res = False
        for i in range(len(board)):
            for j in range(len(board[i])):
                if board[i][j]==word[0]:
                    res |= dfs(i, j, word)
        return res
